cmd/kube-proxy/app/server.go

```diff
@@ -626,28 +634,27 @@ func serveMetrics(bindAddress string, proxyMode kubeproxyconfig.ProxyMode, enabl
 
 	configz.InstallHandler(proxyMux)
 
-	fn := func() {
+	fn := func(ctx context.Context) {
 		err := http.ListenAndServe(bindAddress, proxyMux)
 		if err != nil {
 			err = fmt.Errorf("starting metrics server failed: %v", err)
 			utilruntime.HandleError(err)
 			if errCh != nil {
 				errCh <- err
 				// if in hardfail mode, never retry again
-				blockCh := make(chan error)
-				<-blockCh
+				cancel()
 			}
 		}
 	}
-	go wait.Until(fn, 5*time.Second, wait.NeverStop)
+
+	go wait.UntilWithContext(serveCtx, fn, 5*time.Second)
 }
 
 // Run runs the specified ProxyServer.  This should never exit (unless CleanupAndExit is set).
 // TODO: At the moment, Run() cannot return a nil error, otherwise it's caller will never exit. Update callers of Run to handle nil errors.
-func (s *ProxyServer) Run() error {
+func (s *ProxyServer) Run(ctx context.Context) error {
```

Comment:
This has to be cascade to line 753 too , maybe more

Reply:
yah, grep for `wait.NeverStop`

the most important is here, on this `SyncLoop()`
https://github.com/kerthcet/kubernetes/blob/ca0d4982dd4a0296c76dab6dbf54c685186399bc/pkg/proxy/iptables/proxier.go#L489

Reply:
Replace all the neverStop with ctx.Done() under this file.

Reply:
> yah, grep for `wait.NeverStop`
> 
> the most important is here, on this `SyncLoop()` https://github.com/kerthcet/kubernetes/blob/ca0d4982dd4a0296c76dab6dbf54c685186399bc/pkg/proxy/iptables/proxier.go#L489

This is an interface{} and will impact multi packages, also need to change some logics, I'd like to follow this with another PR.

Reply:
I have to check this more in detail, but at first sight this will keep running the sync loop , however, all the informers will be stopped.

The return <-errCh at the end of the method should not be changed to a select{} adding the context signal?

Reply:
Make sense, when context is done, we should also return.